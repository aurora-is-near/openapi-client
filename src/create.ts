import axios, { AxiosInstance, RawAxiosRequestHeaders } from 'axios';
import { noCase } from 'change-case';
import { createRequestFunction } from './request';
import {
  createRefreshTokenInterceptor,
  createEconnresetInterceptor,
  createResponseDebugInterceptor,
  createRequestDebugInterceptor,
  timeoutInterceptor,
} from './interceptors';
import { serializeQueryParams } from './query';
import { OpenApiClientOptions } from './options';

const DEFAULT_REFRESH_STATUS_CODES = [401];
const DEFAULT_REQUEST_TIMEOUT = 15000;

/**
 * Create the base axios instance.
 */
const createAxiosInstance = (
  {
    baseURL,
    refreshAccessToken,
    onError,
    onClientError,
    onTimeoutError,
    onRequest,
    paramsSerializer = serializeQueryParams,
    userAgent,
    appVersion,
    refreshStatusCodes = DEFAULT_REFRESH_STATUS_CODES,
    timeout = DEFAULT_REQUEST_TIMEOUT,
  }: OpenApiClientOptions,
  version: string,
  title: string,
): AxiosInstance => {
  if (!baseURL) {
    throw new Error('A `baseURL` must be given');
  }

  const headers: RawAxiosRequestHeaders = {
    'Content-Type': 'application/json',
    Accept: `application/vnd.${noCase(title).replace(
      /\s+/g,
      '',
    )}+json; version=${version}`,
  };

  if (userAgent) {
    headers['User-Agent'] = userAgent;
  }

  if (appVersion) {
    headers['X-App-Version'] = appVersion;
  }

  const axiosInstance = axios.create({
    baseURL,
    headers,
    paramsSerializer,
    timeout: typeof timeout === 'number' ? timeout : undefined,
  });

  if (typeof timeout === 'object') {
    axiosInstance.interceptors.request.use(timeoutInterceptor(timeout));
  }

  const refreshTokenInterceptor = createRefreshTokenInterceptor(
    refreshStatusCodes,
    refreshAccessToken,
  );

  const econnresetInterceptor = createEconnresetInterceptor();
  const responseDebugInterceptor = createResponseDebugInterceptor(
    onError,
    onClientError,
    onTimeoutError,
  );

  axiosInstance.interceptors.response.use(
    refreshTokenInterceptor.success,
    refreshTokenInterceptor.error,
  );

  axiosInstance.interceptors.response.use(
    econnresetInterceptor.success,
    econnresetInterceptor.error,
  );

  axiosInstance.interceptors.response.use(
    responseDebugInterceptor.success,
    responseDebugInterceptor.error,
  );

  if (onRequest) {
    const requestDebugInterceptor = createRequestDebugInterceptor(
      paramsSerializer,
      onRequest,
    );

    axiosInstance.interceptors.request.use(
      requestDebugInterceptor.success,
      requestDebugInterceptor.error,
    );
  }

  return axiosInstance;
};

/**
 * Get the function for making API requests for a particular client.
 */
export const getRequestFunction = (
  title: string,
  version: string,
  options: OpenApiClientOptions,
) => {
  const axiosInstance = createAxiosInstance(options, version, title);

  return createRequestFunction(
    axiosInstance,
    options.basePath,
    options.getAccessToken,
    options.refreshAccessToken,
  );
};
