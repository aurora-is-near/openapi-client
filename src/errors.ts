export type ValidationError = {
  property: string;
  constraint: string;
  message: string;
};

type OpenApiClientErrorOptions = {
  type?: string;
  errors?: ValidationError[];
};

export type OpenApiClientTimeoutErrorCode = 'ECONNABORTED' | 'ETIMEDOUT';

type OpenApiClientErrorCode = 'EHTTPCLIENT' | 'EHTTPSERVER';

type ErrorCodes = OpenApiClientTimeoutErrorCode | OpenApiClientErrorCode;

class BaseOpenApiClientError extends Error {
  code: ErrorCodes;

  constructor(message: string, code: ErrorCodes) {
    super(message);

    this.code = code;
  }
}

export class OpenApiClientError extends BaseOpenApiClientError {
  statusCode: number;

  type?: string;

  errors?: ValidationError[];

  isOpenApiClientError = true;

  constructor(
    statusCode: number,
    message: string,
    { errors, type }: OpenApiClientErrorOptions = {},
  ) {
    const code = statusCode >= 500 ? 'EHTTPSERVER' : 'EHTTPCLIENT';

    super(
      `${message}${
        errors?.length
          ? ` ${errors
              ?.map(
                (err) => `${err.message} (${err.property} ${err.constraint})`,
              )
              .join(', ')}`
          : ''
      }`,
      code,
    );

    this.statusCode = statusCode;
    this.name = 'OpenApiClientError';
    this.errors = errors;
    this.type = type;
  }
}

/**
 * Determines whether a value is (probably) an error thrown by the API client.
 */
export const isOpenApiClientError = (
  value: unknown,
): value is OpenApiClientError =>
  typeof value === 'object' && !!value && 'isOpenApiClientError' in value;

export class OpenApiClientTimeoutError extends BaseOpenApiClientError {
  isOpenApiClientTimeoutError = true;

  constructor(message: string, code: OpenApiClientTimeoutErrorCode) {
    super(message, code);

    this.name = 'OpenApiClientTimeoutError';
    this.code = code;
  }
}

/**
 * Determines whether a value is (probably) a timeout error thrown by the API client.
 */
export const isOpenApiClientTimeoutError = (
  value: unknown,
): value is OpenApiClientTimeoutError =>
  typeof value === 'object' &&
  !!value &&
  'isOpenApiClientTimeoutError' in value;
