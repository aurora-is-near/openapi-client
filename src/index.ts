// @ts-ignore
// eslint-disable-next-line import/no-unresolved
export * from './generated';

export {
  isOpenApiClientError,
  OpenApiClientError,
  isOpenApiClientTimeoutError,
  OpenApiClientTimeoutError,
} from './errors';
export { OpenApiClientOptions } from './options';
