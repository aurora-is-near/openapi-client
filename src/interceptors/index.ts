export { createRefreshTokenInterceptor } from './refresh-access-token-interceptor';
export { createEconnresetInterceptor } from './econnreset-interceptor';
export { createResponseDebugInterceptor } from './response-debug-interceptor';
export { createRequestDebugInterceptor } from './request-debug-interceptor';
export { timeoutInterceptor } from './timeout-interceptor';
