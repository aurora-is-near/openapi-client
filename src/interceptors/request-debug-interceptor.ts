import { CustomParamsSerializer, InternalAxiosRequestConfig } from 'axios';

export type OnRequestFunction = ({
  method,
  url,
}: {
  method: string;
  url: URL;
}) => void;

/**
 * Axios interceptor to log requests.
 */
export const createRequestDebugInterceptor = (
  paramsSerializer: CustomParamsSerializer,
  onRequest: OnRequestFunction,
) => ({
  success: (config: InternalAxiosRequestConfig) => {
    const { baseURL, method, url, params } = config;

    if (!url || !method) {
      return config;
    }

    const urlObj = new URL(url, baseURL);

    if (params) {
      urlObj.search = `?${paramsSerializer(params)}`;
    }

    onRequest({
      method: method.toUpperCase(),
      url: urlObj,
    });

    return config;
  },
  error: (error: any) => Promise.reject(error),
});
