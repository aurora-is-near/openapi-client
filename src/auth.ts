import decode from 'jwt-decode';
import { OpenApiClientError } from './errors';

type DecodedToken = {
  role?: string | string[];
  exp: number;
};

export type AccessToken = string | null | undefined;

export type TokenRetrieverFunction = () => AccessToken | Promise<AccessToken>;

/**
 * Get a property from the access token.
 */
const getTokenProperty = (accessToken: string, key: keyof DecodedToken) => {
  try {
    return decode<DecodedToken>(accessToken)[key];
  } catch (err) {
    // Failed to decode token, ignore.
    return null;
  }
};

/**
 * Check if the user has the admin role.
 */
const isUserAdmin = (accessToken: string) => {
  const role = getTokenProperty(accessToken, 'role');
  const adminRole = 'wp-admin';

  if (!role) {
    return false;
  }

  return (
    role === adminRole || (Array.isArray(role) && role.includes(adminRole))
  );
};

/**
 * Check if an access token has expired.
 */
const isTokenExpired = (accessToken: string) => {
  const exp = getTokenProperty(accessToken, 'exp');
  const timeNow = new Date().getTime() / 1000;

  if (typeof exp !== 'number') {
    return true;
  }

  return exp < timeNow;
};

/**
 * Throw a 401 Unauthorized error.
 */
const throwUnauthorizedError = (message: string, url: string) => {
  throw new OpenApiClientError(401, `${message} <${url}>`);
};

function assertDefined<T>(
  errorMessage: string,
  url: string,
  value?: T | null,
): asserts value is T {
  if (!value) {
    throwUnauthorizedError(errorMessage, url);
  }
}

const getRefreshedAccessToken = async (
  url: string,
  refreshAccessToken?: TokenRetrieverFunction,
): Promise<string> => {
  assertDefined(
    'Authorization is required but there is no valid access token and no `refreshAccessToken()` function was provided',
    url,
    refreshAccessToken,
  );

  const refreshedToken = await refreshAccessToken();

  assertDefined(
    'Authorization is required but there is no valid access token and nothing was returned from `refreshAccessToken()`',
    url,
    refreshedToken,
  );

  return refreshedToken;
};

/**
 * Get the bearer authorization header.
 */
export const getAuthorizationHeader = async (
  url: string,
  secure: boolean,
  getAccessToken?: TokenRetrieverFunction,
  refreshAccessToken?: TokenRetrieverFunction,
) => {
  if (secure) {
    assertDefined(
      'Authorization is required but no `getAccessToken()` function was provided',
      url,
      getAccessToken,
    );
  }

  const noop = () => null;
  let accessToken = await (getAccessToken ?? noop)();

  if (!secure && (!accessToken || !isUserAdmin(accessToken))) {
    return null;
  }

  if (!accessToken) {
    accessToken = await getRefreshedAccessToken(url, refreshAccessToken);
  }

  // If the user is an admin and their token has expired but the route was not
  // secure anyway then we can still go ahead and make the request.
  if (!secure && isTokenExpired(accessToken)) {
    return null;
  }

  if (isTokenExpired(accessToken)) {
    accessToken = await getRefreshedAccessToken(url, refreshAccessToken);

    if (isTokenExpired(accessToken)) {
      throwUnauthorizedError(
        'Authorization is required but the access token has expired and `refreshAccessToken()` also returned an expired token',
        url,
      );
    }
  }

  return `Bearer ${accessToken}`;
};
