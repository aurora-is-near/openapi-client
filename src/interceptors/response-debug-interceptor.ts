import { AxiosResponse } from 'axios';
import { OpenApiClientError, OpenApiClientTimeoutError } from '../errors';

export const createResponseDebugInterceptor = (
  onError?: (error: any) => void,
  onClientError?: (error: any) => void,
  onTimeoutError?: (error: any) => void,
) => ({
  success: (res: AxiosResponse) => res,
  error: (error: any) => {
    const { config, response, message, statusCode, code } = error;

    const { method } = config;
    const { data, status } = response || {};
    const finalStatus = status ?? statusCode;

    const isFiniteStatus = Number.isFinite(finalStatus);
    const isServerError = isFiniteStatus && finalStatus >= 500;
    const isClientError =
      isFiniteStatus && finalStatus >= 400 && finalStatus < 500;
    const isTimeoutError = code === 'ETIMEDOUT' || code === 'ECONNABORTED';

    const endpoint = `/${config.url}`.replace(/\/\//g, '/');
    const url = `${config.baseURL}${endpoint}`;
    const msg = `${finalStatus} ${
      data?.message || message
    } <${method.toUpperCase()} ${url}>`;

    const openApiClientError = new OpenApiClientError(finalStatus, msg, {
      type: data?.type,
      errors: data?.errors,
    });

    const logError = onError || console.error;
    const logClientError = onClientError || console.warn;
    const logTimeoutError = onTimeoutError || console.warn;

    if (isTimeoutError) {
      // Axios returns the timeout as code NOT statusCode
      const openApiTimeoutError = new OpenApiClientTimeoutError(message, code);

      logTimeoutError(openApiTimeoutError);

      throw openApiTimeoutError;
    }

    if (isServerError) {
      logError(openApiClientError);
    }

    if (isClientError) {
      logClientError(openApiClientError);
    }

    throw openApiClientError;
  },
});
