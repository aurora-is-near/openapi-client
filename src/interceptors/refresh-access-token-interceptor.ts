import axios, { AxiosError, AxiosResponse } from 'axios';
import { TokenRetrieverFunction } from '../auth';

declare module 'axios' {
  export interface InternalAxiosRequestConfig {
    _retry?: boolean;
  }
}

/**
 * Axios interceptor to retry a request after refreshing an access token.
 */
export const createRefreshTokenInterceptor = (
  refreshStatusCodes: number[],
  refreshAccessToken?: TokenRetrieverFunction,
) => ({
  success: (res: AxiosResponse) => res,
  error: async (error: AxiosError) => {
    const { config, response } = error;
    const { status } = response ?? {};

    if (
      config &&
      refreshAccessToken &&
      status &&
      refreshStatusCodes.includes(status) && // Authentication failed for original req
      config.headers?.Authorization && // and it included an Authorization header
      !config._retry // and we haven't already retried
    ) {
      config._retry = true;

      const token = await refreshAccessToken();

      if (!token) {
        return Promise.reject(error);
      }

      config.headers.Authorization = `Bearer ${token}`;

      return axios(config);
    }

    return Promise.reject(error);
  },
});
