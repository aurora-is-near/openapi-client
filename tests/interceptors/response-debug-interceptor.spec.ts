import { AxiosError, AxiosResponse } from 'axios';
import {
  OpenApiClientError,
  OpenApiClientTimeoutError,
} from '../../src/errors';
import { createResponseDebugInterceptor } from '../../src/interceptors';

const originalConsoleWarn = console.warn;
const originalConsoleError = console.error;

describe('createResponseDebugInterceptor', () => {
  beforeEach(() => {
    console.warn = originalConsoleWarn;
    console.error = originalConsoleError;
  });

  describe('error', () => {
    it('logs a 500 error', async () => {
      console.error = jest.fn();

      const error = {
        response: {
          status: 500,
          data: {
            statusCode: 500,
            message: 'Internal Server Error',
            name: 'BadRequest',
            errors: [],
          },
        },
        config: {
          url: '/endpoint',
          baseURL: 'http://api.com',
          method: 'get',
        },
      } as unknown as AxiosError;

      const interceptor = createResponseDebugInterceptor();

      let err;

      try {
        interceptor.error(error);
      } catch (e) {
        err = e;
      }

      expect((err as OpenApiClientError).message).toBe(
        '500 Internal Server Error <GET http://api.com/endpoint>',
      );

      expect((err as OpenApiClientError).statusCode).toBe(500);
      expect((err as OpenApiClientError).name).toBe('OpenApiClientError');
      expect((err as OpenApiClientError).errors).toEqual([]);

      expect(console.error).toHaveBeenCalledTimes(1);
      expect((console.error as jest.Mock).mock.calls[0][0]).toMatchSnapshot();
    });

    it('logs a 400 error as a warning', async () => {
      console.warn = jest.fn();

      const errors = [
        {
          constraint: 'isInt',
          message: 'id must be an integer number',
          property: 'id',
        },
        {
          constraint: 'isString',
          message: 'name must be a string',
          property: 'name',
        },
      ];

      const error = {
        response: {
          status: 400,
          data: {
            statusCode: 400,
            message: 'Bad Request',
            name: 'BadRequest',
            type: '/probs/the-thing-expired',
            errors,
          },
        },
        config: {
          url: '/endpoint',
          baseURL: 'http://api.com',
          method: 'put',
        },
      } as unknown as AxiosError;

      const interceptor = createResponseDebugInterceptor();

      let err;

      try {
        interceptor.error(error);
      } catch (e) {
        err = e;
      }

      expect((err as OpenApiClientError).message).toBe(
        '400 Bad Request <PUT http://api.com/endpoint> id must be an integer number (id isInt), name must be a string (name isString)',
      );

      expect((err as OpenApiClientError).statusCode).toBe(400);
      expect((err as OpenApiClientError).name).toBe('OpenApiClientError');
      expect((err as OpenApiClientError).errors).toEqual(errors);
      expect((err as OpenApiClientError).type).toBe('/probs/the-thing-expired');

      expect(console.warn).toHaveBeenCalledTimes(1);
      expect((console.warn as jest.Mock).mock.calls[0][0]).toMatchSnapshot();
    });

    it('logs a 500 error using a custom error handler', async () => {
      console.error = jest.fn();

      const error = {
        response: {
          status: 500,
          data: {
            statusCode: 500,
            message: 'Internal Server Error',
            name: 'BadRequest',
            errors: [],
          },
        },
        config: {
          url: '/endpoint',
          baseURL: 'http://api.com',
          method: 'post',
        },
      } as unknown as AxiosError;

      const onError = jest.fn();
      const interceptor = createResponseDebugInterceptor(onError);

      let err;

      try {
        interceptor.error(error);
      } catch (e) {
        err = e;
      }

      expect((err as OpenApiClientError).message).toBe(
        '500 Internal Server Error <POST http://api.com/endpoint>',
      );

      expect((err as OpenApiClientError).statusCode).toBe(500);
      expect((err as OpenApiClientError).name).toBe('OpenApiClientError');
      expect((err as OpenApiClientError).errors).toEqual([]);

      expect(onError).toHaveBeenCalledTimes(1);
      expect(onError).toHaveBeenCalledWith(err);

      expect(console.error).not.toHaveBeenCalled();
    });

    it.each([undefined, {}, ''])(
      'logs an error when the API response data is "%s"',
      async (data) => {
        console.error = jest.fn();

        const error = {
          message: 'Bad thing',
          response: {
            status: 500,
            data,
          },
          config: {
            url: '/endpoint',
            baseURL: 'http://api.com',
            method: 'post',
          },
        } as unknown as AxiosError;

        const interceptor = createResponseDebugInterceptor();

        let err;

        try {
          interceptor.error(error);
        } catch (e) {
          err = e;
        }

        expect((err as OpenApiClientError).message).toBe(
          '500 Bad thing <POST http://api.com/endpoint>',
        );

        expect((err as OpenApiClientError).statusCode).toBe(500);
        expect((err as OpenApiClientError).name).toBe('OpenApiClientError');

        expect(console.error).toHaveBeenCalledTimes(1);
        expect((console.error as jest.Mock).mock.calls[0][0]).toMatchSnapshot();
      },
    );

    it('logs an error when the response body does not exist', async () => {
      console.error = jest.fn();

      const error = {
        statusCode: 500,
        message: 'Bad thing',
        config: {
          url: '/endpoint',
          baseURL: 'http://api.com',
          method: 'get',
        },
      } as unknown as AxiosError;

      const interceptor = createResponseDebugInterceptor();

      let err;

      try {
        interceptor.error(error);
      } catch (e) {
        err = e;
      }

      expect((err as OpenApiClientError).message).toBe(
        '500 Bad thing <GET http://api.com/endpoint>',
      );

      expect((err as OpenApiClientError).statusCode).toBe(500);
      expect((err as OpenApiClientError).name).toBe('OpenApiClientError');
      expect((err as OpenApiClientError).errors).toBeUndefined();

      expect(console.error).toHaveBeenCalledTimes(1);
      expect((console.error as jest.Mock).mock.calls[0][0]).toMatchSnapshot();
    });

    it.each(['ECONNABORTED', 'ETIMEDOUT'])(
      'logs a time out error as a warning when the code is "%s"',
      async (code) => {
        console.warn = jest.fn();

        const error = {
          code,
          message: 'Request aborted',
          config: {
            url: '/endpoint',
            baseURL: 'http://api.com',
            method: 'get',
          },
        } as unknown as AxiosError;

        const interceptor = createResponseDebugInterceptor();

        let err;

        try {
          interceptor.error(error);
        } catch (e) {
          err = e;
        }

        expect((err as OpenApiClientTimeoutError).message).toBe(
          'Request aborted',
        );

        expect((err as OpenApiClientTimeoutError).code).toBe(code);
        expect((err as OpenApiClientTimeoutError).name).toBe(
          'OpenApiClientTimeoutError',
        );

        expect(console.warn).toHaveBeenCalledTimes(1);
        expect((console.warn as jest.Mock).mock.calls[0][0]).toMatchSnapshot();
      },
    );

    it.each(['ECONNABORTED', 'ETIMEDOUT'])(
      'logs a timeout error using a custom timeout error handler when the code is "%s"',
      async (code) => {
        console.warn = jest.fn();

        const error = {
          code,
          message: 'Request aborted',
          config: {
            url: '/endpoint',
            baseURL: 'http://api.com',
            method: 'get',
          },
        } as unknown as AxiosError;

        const onClientError = jest.fn();
        const onError = jest.fn();
        const onTimeoutError = jest.fn();
        const interceptor = createResponseDebugInterceptor(
          onError,
          onClientError,
          onTimeoutError,
        );

        let err;

        try {
          interceptor.error(error);
        } catch (e) {
          err = e;
        }

        expect((err as OpenApiClientTimeoutError).message).toBe(
          'Request aborted',
        );

        expect((err as OpenApiClientTimeoutError).code).toBe(code);
        expect((err as OpenApiClientTimeoutError).name).toBe(
          'OpenApiClientTimeoutError',
        );

        expect(console.warn).not.toHaveBeenCalled();
        expect(onTimeoutError).toHaveBeenCalledTimes(1);
        expect(onTimeoutError).toHaveBeenCalledWith(err);
        expect(onClientError).not.toHaveBeenCalled();
        expect(onError).not.toHaveBeenCalled();
      },
    );
  });

  describe('success', () => {
    it('returns the response on success', () => {
      const mockResponse = { foo: 'bar' };
      const interceptor = createResponseDebugInterceptor();

      const result = interceptor.success(
        mockResponse as unknown as AxiosResponse,
      );

      expect(result).toEqual(mockResponse);
    });
  });
});
