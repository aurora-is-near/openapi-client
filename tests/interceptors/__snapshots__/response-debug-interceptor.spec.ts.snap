// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createResponseDebugInterceptor error logs a 400 error as a warning 1`] = `[OpenApiClientError: 400 Bad Request <PUT http://api.com/endpoint> id must be an integer number (id isInt), name must be a string (name isString)]`;

exports[`createResponseDebugInterceptor error logs a 500 error 1`] = `[OpenApiClientError: 500 Internal Server Error <GET http://api.com/endpoint>]`;

exports[`createResponseDebugInterceptor error logs a time out error as a warning when the code is "ECONNABORTED" 1`] = `[OpenApiClientTimeoutError: Request aborted]`;

exports[`createResponseDebugInterceptor error logs a time out error as a warning when the code is "ETIMEDOUT" 1`] = `[OpenApiClientTimeoutError: Request aborted]`;

exports[`createResponseDebugInterceptor error logs an error when the API response data is "" 1`] = `[OpenApiClientError: 500 Bad thing <POST http://api.com/endpoint>]`;

exports[`createResponseDebugInterceptor error logs an error when the API response data is "{}" 1`] = `[OpenApiClientError: 500 Bad thing <POST http://api.com/endpoint>]`;

exports[`createResponseDebugInterceptor error logs an error when the API response data is "undefined" 1`] = `[OpenApiClientError: 500 Bad thing <POST http://api.com/endpoint>]`;

exports[`createResponseDebugInterceptor error logs an error when the response body does not exist 1`] = `[OpenApiClientError: 500 Bad thing <GET http://api.com/endpoint>]`;
